File: DP/leetcode120.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        # tri_len = len(triangle) - 1
        # ans = [0] * len(triangle[tri_len])
        ans = [0] * (len(triangle) + 1)        
        for i in triangle[::-1]:
            for j in range(len(i)):
                ans[j] = i[j] + min(ans[j], ans[j + 1])
        
        return ans[0]
                

File: DP/leetcode120.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int[][] minCost = new int[triangle.size() + 1][triangle.size() + 1];
        for (int i = triangle.size() - 1; i >= 0; i --){
            int[] tempList = new int[triangle.size()];
            for (int j = 0; j < triangle.get(i).size(); j ++){
                tempList[j] = Math.min((triangle.get(i).get(j) + minCost[i + 1][j]), (triangle.get(i).get(j) + minCost[i + 1][j + 1]));
            }
            minCost[i] = tempList;
        }
        return minCost[0][0];
        
    }
}

// Should be easy. But java adds the complexity.